AST visitor generation does not work when the node and visitor are defined in different projects

**Describe the bug**
If a syntax tree is defined in some project A like this:
```cs
[SyntaxTree]
public abstract partial record AstNode { ... }
```
and then a visitor is defined in a project B like this:
```cs
[SyntaxTreeVisitor(typeof(AstNode))
public abstract partial class SomeAstVisitor { ... }
```
The syntax tree generation will not work. No error is produced from the source generators and no exception is indicated by CSC, as if the generation completed successfully, but later during build time errors arise because the `Visit` functions (or whatever name the user expects, in case they are using the custom method name feature) do not exist, this is, the generator has silently not generated any methods.
```cs
error CS0115: 'SomeAstVisitor.Visit(SomeNodeType)': no suitable method found to override
```

This only happens when the tree node (`AstNode` in project A) definition and the visitor definition (`SomeAstVisitor` in project B) are in different C# projects. 

Regarding package installation, the following configurations have been tried:
- **Configuration 1**
  - `Yoakke.SyntaxTree` in both projects A and B.
  - `Yoakke.SyntaxTree.Generators` in both projects A and B.
- **Configuration 2**
  - `Yoakke.SyntaxTree` in project A only.
  - `Yoakke.SyntaxTree.Generators` in project B only.
- **Configuration 3**
  - `Yoakke.SyntaxTree` in both projects A and B.
  - `Yoakke.SyntaxTree.Generators` in project B only.
- **Configuration 4**
  - `Yoakke.SyntaxTree` in project A only.
  - `Yoakke.SyntaxTree.Generators` in both projects A and B.

In all configurations the error is exactly the same, so it seems unlikely this is the cause.

**Which libraries does it affect?**
- `Yoakke.SyntaxTree`
- `Yoakke.SyntaxTree.Generators`

**To Reproduce**
1. Write a simple compiler with an AST and some visitor, of any complexity, in a project "A".
2. Move the AST definitions to a separate C# Class Library project "B".
3. Add a reference from your current project "A" to the new project "B", and the necessary `using` directives for it to compile.
4. The visitor should use some or override some call to `Visit` (or the selected method name) to make it evident if the code generation has provided any implementation.
5. Build the project. No warnings related to the source generator `SyntaxTreeSourceGenerator` should appear.
6. Observe an error in every call to `Visit` (or the selected method name) because no definition of that method exists.

**Expected behavior**
The code should build and have the same exact behavior as if the definitions were in the same project.

**Environment (please complete the following information):**
- OS: 10.0.19043
- .NET version: 5.0.8 (35964c9215), SDK 5.0.302
- Version: 2021.8.11-11.54.18-nightly

--- a/Sources/Core/Yoakke.SourceGenerator.Common/RoslynExtensions/SymbolExtensions.cs
+++ b/Sources/Core/Yoakke.SourceGenerator.Common/RoslynExtensions/SymbolExtensions.cs
@@ -252,5 +252,46 @@ namespace Yoakke.SourceGenerator.Common.RoslynExtensions
                 .FirstOrDefault(attr => SymbolEqualityComparer.Default.Equals(attr.AttributeClass, attributeSymbol));
             return attributeData is not null;
         }
+
+        /// <summary>
+        /// Retrieves all <see cref="INamedTypeSymbol"/>s defined in the given <see cref="IAssemblySymbol"/>.
+        /// </summary>
+        /// <param name="assembly">The <see cref="IAssemblySymbol"/> to search for types.</param>
+        /// <returns>The sequence of declared types in <paramref name="assembly"/>.</returns>
+        public static IEnumerable<INamedTypeSymbol> GetAllDeclaredTypes(this IAssemblySymbol assembly) =>
+            assembly.GlobalNamespace.GetAllDeclaredTypes();
+
+        /// <summary>
+        /// Retrieves all <see cref="INamedTypeSymbol"/>s defined in the given <see cref="INamespaceSymbol"/>.
+        /// </summary>
+        /// <param name="namespace">The <see cref="INamespaceSymbol"/> to search for types.</param>
+        /// <returns>The sequence of declared types in <paramref name="namespace"/>.</returns>
+        public static IEnumerable<INamedTypeSymbol> GetAllDeclaredTypes(this INamespaceSymbol @namespace)
+        {
+            foreach (var subNs in @namespace.GetNamespaceMembers())
+            {
+                foreach (var type in subNs.GetAllDeclaredTypes()) yield return type;
+            }
+
+            foreach (var type in @namespace.GetTypeMembers())
+            {
+                yield return type;
+                foreach (var subType in type.GetAllDeclaredTypes()) yield return subType;
+            }
+        }
+
+        /// <summary>
+        /// Retrieves all <see cref="INamedTypeSymbol"/>s defined in the given <see cref="INamedTypeSymbol"/>.
+        /// </summary>
+        /// <param name="type">The <see cref="INamedTypeSymbol"/> to search for types.</param>
+        /// <returns>The sequence of declared types in <paramref name="type"/>.</returns>
+        public static IEnumerable<INamedTypeSymbol> GetAllDeclaredTypes(this INamedTypeSymbol type)
+        {
+            foreach (var subType in type.GetTypeMembers())
+            {
+                yield return subType;
+                foreach (var subSubType in subType.GetAllDeclaredTypes()) yield return subSubType;
+            }
+        }
     }
 }

--- a/Sources/Core/Yoakke.SyntaxTree.Generator/SyntaxTreeSourceGenerator.cs
+++ b/Sources/Core/Yoakke.SyntaxTree.Generator/SyntaxTreeSourceGenerator.cs
@@ -99,7 +99,7 @@ namespace Yoakke.SyntaxTree.Generator
                 var visitor = new Visitor(symbol!);
                 foreach (var attr in attrs)
                 {
-                    visitor.Overrides[attr.NodeType!] = new()
+                    visitor.Overrides[attr.NodeType!] = new(attr.NodeType!)
                     {
                         MethodName = attr.MethodName,
                         ReturnType = attr.ReturnType,
@@ -108,8 +108,11 @@ namespace Yoakke.SyntaxTree.Generator
                 visitors.Add(visitor);
             }
 
+            // We recollect all nodes the visitors reference, as those can be in different assemblies
+            var visitableNodes = this.CollectVisitorReferencedNodes(visitors);
+
             // For the visitors we build a node hierarchy
-            var rootNodes = BuildNodeHierarchy(nodes);
+            var rootNodes = BuildNodeHierarchy(visitableNodes);
             // We fill up the hierarchy with visitor information
             foreach (var visitor in visitors)
             {
@@ -282,7 +285,7 @@ namespace Yoakke.SyntaxTree.Generator
             if (visitor.Overrides.TryGetValue(currentNode.NodeClass, out var newOverride))
             {
                 // We override by filling in anything new that has been provided
-                currentOverride = new()
+                currentOverride = new(currentNode.NodeClass)
                 {
                     MethodName = newOverride.MethodName ?? currentOverride?.MethodName ?? "Visit",
                     ReturnType = newOverride.ReturnType ?? currentOverride?.ReturnType ?? voidType,
@@ -294,6 +297,33 @@ namespace Yoakke.SyntaxTree.Generator
 
             // Visit children
             foreach (var child in currentNode.Children) this.PopulateNodeWithVisitorOverrides(visitor, currentOverride, child);
+        }
+
+        private HashSet<INamedTypeSymbol> CollectVisitorReferencedNodes(List<Visitor> visitors)
+        {
+            // NOTE: False positive
+#pragma warning disable RS1024 // Compare symbols correctly
+            var visitableNodes = new HashSet<INamedTypeSymbol>(SymbolEqualityComparer.Default);
+            var visitedAssemblys = new HashSet<IAssemblySymbol>(SymbolEqualityComparer.Default);
+#pragma warning restore RS1024 // Compare symbols correctly
+            foreach (var visitor in visitors)
+            {
+                foreach (var nodeClass in visitor.Overrides.Values.Select(o => o.NodeClass))
+                {
+                    // If the assembly of the node is already visited, skip the recursive check
+                    if (!visitedAssemblys.Add(nodeClass.ContainingAssembly)) continue;
+
+                    // New, unvisited assembly, filter out relevant nodes
+                    var typesInAssembly = nodeClass.ContainingAssembly.GetAllDeclaredTypes();
+                    foreach (var symbol in typesInAssembly)
+                    {
+                        if (!this.IsSyntaxTreeNode(symbol)) continue;
+                        // NOTE: Here we don't require it to be declarable inside, since the visitor is completely external
+                        visitableNodes.Add(symbol);
+                    }
+                }
+            }
+            return visitableNodes;
         }
 
         private static List<MetaNode> BuildNodeHierarchy(HashSet<INamedTypeSymbol> symbols)

--- a/Sources/Core/Yoakke.SyntaxTree.Generator/VisitorOverride.cs
+++ b/Sources/Core/Yoakke.SyntaxTree.Generator/VisitorOverride.cs
@@ -15,6 +15,11 @@ namespace Yoakke.SyntaxTree.Generator
     internal class VisitorOverride
     {
         /// <summary>
+        /// The node class this override applies to.
+        /// </summary>
+        public INamedTypeSymbol NodeClass { get; }
+
+        /// <summary>
         /// Override for the method name.
         /// </summary>
         public string? MethodName { get; set; }
@@ -23,5 +28,14 @@ namespace Yoakke.SyntaxTree.Generator
         /// Override for the method return type.
         /// </summary>
         public INamedTypeSymbol? ReturnType { get; set; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="VisitorOverride"/> class.
+        /// </summary>
+        /// <param name="nodeClass">The node class this override applies to.</param>
+        public VisitorOverride(INamedTypeSymbol nodeClass)
+        {
+            this.NodeClass = nodeClass;
+        }
     }
 }

--- a/Sources/Tests/Yoakke.SyntaxTree.Tests.Tree/ExternalAst.cs
+++ b/Sources/Tests/Yoakke.SyntaxTree.Tests.Tree/ExternalAst.cs
@@ -1,0 +1,15 @@
+using System;
+using Yoakke.SyntaxTree.Attributes;
+
+namespace Yoakke.SyntaxTree.Tests.Tree
+{
+    [SyntaxTree]
+    public abstract partial record ExternalAst
+    {
+        public partial record Foo : ExternalAst;
+
+        public partial record Bar : ExternalAst;
+
+        public partial record Baz : ExternalAst;
+    }
+}

--- a/Sources/Tests/Yoakke.SyntaxTree.Tests.Tree/Yoakke.SyntaxTree.Tests.Tree.csproj
+++ b/Sources/Tests/Yoakke.SyntaxTree.Tests.Tree/Yoakke.SyntaxTree.Tests.Tree.csproj
@@ -1,0 +1,12 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net5.0</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="$(SolutionDir)\Core\Yoakke.SyntaxTree\Yoakke.SyntaxTree.csproj" />
+    <ProjectReference Include="$(SolutionDir)\Core\Yoakke.SyntaxTree.Generator\Yoakke.SyntaxTree.Generator.csproj" OutputItemType="Analyzer" ReferenceOutputAssembly="False" />
+  </ItemGroup>
+
+</Project>

--- a/Sources/Tests/Yoakke.SyntaxTree.Tests/Issue72Tests.cs
+++ b/Sources/Tests/Yoakke.SyntaxTree.Tests/Issue72Tests.cs
@@ -1,0 +1,61 @@
+// Copyright (c) 2021 Yoakke.
+// Licensed under the Apache License, Version 2.0.
+// Source repository: https://github.com/LanguageDev/Yoakke
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Yoakke.SyntaxTree.Attributes;
+using Yoakke.SyntaxTree.Tests.Tree;
+
+namespace Yoakke.SyntaxTree.Tests
+{
+    // https://github.com/LanguageDev/Yoakke/issues/72
+    [TestClass]
+    public partial class Issue72Tests
+    {
+        [SyntaxTreeVisitor(typeof(ExternalAst), ReturnType = typeof(string))]
+        internal abstract partial class ExternalVisitorBase
+        {
+        }
+
+        internal partial class ExternalVisitor : ExternalVisitorBase
+        {
+            public string Stringify(ExternalAst ast) => this.Visit(ast);
+
+            protected override string Visit(ExternalAst.Foo foo) => "got a foo";
+
+            protected override string Visit(ExternalAst.Bar bar) => "got a bar";
+
+            protected override string Visit(ExternalAst.Baz baz) => "got a baz";
+        }
+
+        [TestMethod]
+        public void TestVisitingFoo()
+        {
+            var visitor = new ExternalVisitor();
+            ExternalAst ast = new ExternalAst.Foo();
+            Assert.AreEqual("got a foo", visitor.Stringify(ast));
+        }
+
+        [TestMethod]
+        public void TestVisitingBar()
+        {
+            var visitor = new ExternalVisitor();
+            ExternalAst ast = new ExternalAst.Bar();
+            Assert.AreEqual("got a bar", visitor.Stringify(ast));
+        }
+
+        [TestMethod]
+        public void TestVisitingBaz()
+        {
+            var visitor = new ExternalVisitor();
+            ExternalAst ast = new ExternalAst.Baz();
+            Assert.AreEqual("got a baz", visitor.Stringify(ast));
+        }
+    }
+}

--- a/Sources/Tests/Yoakke.SyntaxTree.Tests/Yoakke.SyntaxTree.Tests.csproj
+++ b/Sources/Tests/Yoakke.SyntaxTree.Tests/Yoakke.SyntaxTree.Tests.csproj
@@ -23,6 +23,7 @@
   <ItemGroup>
     <ProjectReference Include="$(SolutionDir)\Core\Yoakke.SyntaxTree\Yoakke.SyntaxTree.csproj" />
     <ProjectReference Include="$(SolutionDir)\Core\Yoakke.SyntaxTree.Generator\Yoakke.SyntaxTree.Generator.csproj" OutputItemType="Analyzer" ReferenceOutputAssembly="False" />
+    <ProjectReference Include="$(SolutionDir)\Tests\Yoakke.SyntaxTree.Tests.Tree\Yoakke.SyntaxTree.Tests.Tree.csproj" />
   </ItemGroup>
 
 </Project>

--- a/Sources/Yoakke.sln
+++ b/Sources/Yoakke.sln
@@ -100,7 +100,9 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Yoakke.Ir.Sample", "Example
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Platforms", "Platforms", "{A482C708-3225-47F3-8DE3-BB354888CC51}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Yoakke.NugetMaintainer", "Tools\Yoakke.NugetMaintainer\Yoakke.NugetMaintainer.csproj", "{7BD28C2F-18D6-4C3A-9899-469775B001D8}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Yoakke.NugetMaintainer", "Tools\Yoakke.NugetMaintainer\Yoakke.NugetMaintainer.csproj", "{7BD28C2F-18D6-4C3A-9899-469775B001D8}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Yoakke.SyntaxTree.Tests.Tree", "Tests\Yoakke.SyntaxTree.Tests.Tree\Yoakke.SyntaxTree.Tests.Tree.csproj", "{2E1E2A07-9EC8-462D-B001-8AB44F973286}"
 EndProject
 Global
 	GlobalSection(SolutionConfigurationPlatforms) = preSolution
@@ -252,6 +254,10 @@ Global
 		{7BD28C2F-18D6-4C3A-9899-469775B001D8}.Debug|Any CPU.Build.0 = Debug|Any CPU
 		{7BD28C2F-18D6-4C3A-9899-469775B001D8}.Release|Any CPU.ActiveCfg = Release|Any CPU
 		{7BD28C2F-18D6-4C3A-9899-469775B001D8}.Release|Any CPU.Build.0 = Release|Any CPU
+		{2E1E2A07-9EC8-462D-B001-8AB44F973286}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
+		{2E1E2A07-9EC8-462D-B001-8AB44F973286}.Debug|Any CPU.Build.0 = Debug|Any CPU
+		{2E1E2A07-9EC8-462D-B001-8AB44F973286}.Release|Any CPU.ActiveCfg = Release|Any CPU
+		{2E1E2A07-9EC8-462D-B001-8AB44F973286}.Release|Any CPU.Build.0 = Release|Any CPU
 	EndGlobalSection
 	GlobalSection(SolutionProperties) = preSolution
 		HideSolutionNode = FALSE
@@ -284,6 +290,7 @@ Global
 		{04888C13-EA59-4AC4-9DFE-BEE103AFA006} = {58417641-179E-4F53-9528-234B4E87944E}
 		{D9A26C9B-EB6D-4EDB-923B-ED529E4C5A8D} = {B042CA2F-FAD5-4148-89C8-3310AF17BC14}
 		{B5C7DD29-4D2C-46B7-B940-FA5C2680C382} = {B042CA2F-FAD5-4148-89C8-3310AF17BC14}
+		{2E1E2A07-9EC8-462D-B001-8AB44F973286} = {58417641-179E-4F53-9528-234B4E87944E}
 	EndGlobalSection
 	GlobalSection(ExtensibilityGlobals) = postSolution
 		SolutionGuid = {1EDE8F07-AE3A-4556-AEF0-3AC518F18F92}

